website/integrations: add bitwarden (#14922)
goauthentik/authentik: website/integrations/services/bitwarden/index.mdx

---
title: Integrate with Bitwarden
sidebar_label: Bitwarden
support_level: community
---

import TabItem from "@theme/TabItem";
import Tabs from "@theme/Tabs";

## What is Bitwarden

> 

Reviewer: @dewi-tik what do you think about adding this to the "Dashboards" section of our integration guides, instead of Miscellaneous? We want to minimize what is under the Miscellaneous section whenever possible, and there is something called Linkwarden already under "Dashboard"...

> Bitwarden is a freemium, open-source password management service that helps users store, manage, and share sensitive information securely online.
>
> -- https://bitwarden.com/

## Preparation

The following placeholders are used in this guide:

- `authentik.company` is the FQDN of the authentik installation.
- `bitwarden.company` is the FQDN of the Bitwarden installation (if using self-hosted Bitwarden).

:::note
This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

## Configuration methods

> You can configure Bitwarden to use either OIDC or SAML, and this guide explains both options.

Reviewer 2: ```suggestion
You can configure Bitwarden to use either OIDC or SAML; this guide explains both options.
```

<Tabs
  defaultValue="oidc"
  values={[
    { label: "Log in with OIDC", value: "oidc" },
    { label: "Log in with SAML", value: "saml" },
  ]}>
  <TabItem value="oidc">

## authentik configuration

To support the integration of Bitwarden with authentik, you need to create a property mapping, and an application/provider pair in authentik.

### Create a scope mapping

> Bitwarden requires a first and last name for each user. By default authentik only provides a full name, as a single string value. Therefore you need to create a property mapping to provide first and last names to Bitwarden.

Reviewer 2: ```suggestion
Bitwarden requires a first and last name for every user. However, authentik, by default, only supplies a full name as a single string. As a result, a property mapping must be created to separate and provide first and last names to Bitwarden.
```

1. Log in to authentik as an administrator, and open the authentik Admin interface.
2. Navigate to **Customization** > **Property Mappings** and click **Create**.

>     - **Select type**: select **Scope Mapping** as the property mapping type.

Reviewer 2: ```suggestion
    - **Select type**: select **Scope Mapping**.
```

    - **Configure the Scope Mapping**: Provide a descriptive name (e.g. `Bitwarden Profile Scope`), and an optional description.

        - **Scope name**: `profile`
        - **Expression**:

>         ```python showLineNumbers

Reviewer 2: Should we get someone to look over this python or can we confirm it's fine as-is?

Author: I've tested it and it's working in a few OIDC setups. So should be ok

            "groups": [group.name for group in request.user.ak_groups.all()],
            "surname": request.user.name.rsplit(" ", 1)[-1],
            "givenname": request.user.name.rsplit(" ", 1)[0],
        }
        ```

3. Click **Finish** to save the property mapping.

### Create an application and provider in authentik

1. Log in to authentik as an administrator, and open the authentik Admin interface.
2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **OAuth2/OpenID Connect** as the provider type.
    - **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
        - Note the **Client ID**, **Client Secret**, and **slug** values because they will be required later.

>         - Set a `Strict` redirect URI to `https://sso.bitwarden.com/oidc-signin`.

Reviewer 2: i'd do "set a strict redirect uri to x for bitwarden SAAS, y for bitwarden saas (eu), or z for a self-managed installation"

what do you think? However, it feels a little long so your original implementation could be modified to put each redirect URI on a different line without necessarily defaulting to the normal version as well as specifying its the SAAS version

Author: What do you think about this new version?

Reviewer 2: Looks perfect!

            - or `https://sso.bitwarden.eu/oidc-signin` for EU based Bitwarden accounts.
            - or `https://bitwarden.company/oidc-signin` for self-hosted Bitwarden.
        - Select any available signing key.
        - Under **Advanced protocol settings**:
            - Remove `authentik default OAuth Mapping: OpenID 'profile'` from **Selected Scopes**.

>             - Add the scope mapping that you created (e.g. `Bitwarden Profile Scope`) to **Selected Scopes**.

Reviewer 2: I wouldn't keep the e.g. one should know the thing they just created and adding an example here feels a little weird. what do you think?

    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

## Bitwarden configuration

1. Log in to the [Bitwarden dashboard](https://vault.bitwarden.com/#/login) as an administrator (select `Accessing: bitwarden.eu` for EU based Bitwarden accounts). For self-hosted Bitwarden use `https://bitwarden.company/#/login`.
2. In the sidebar, navigate to **Admin Console** > **Settings** > **Single sign-on**, and enter the following settings:

    - **Allow SSO authentication**: Checked.
    - **SSO Identifier**: enter a globally unique SSO identifier (this is not required if using self-hosted Bitwarden, or if you have claimed a domain, see the [Bitwarden Claimed Domains documentation](https://bitwarden.com/help/claimed-domains/)).
    - **Type**: `OIDC`
    - Under **OpenID connect configuration**:

>         - **Authority**: `https://authentik.company/application/o/<application-slug>/`

Reviewer 2: Tana actually prefers underscores for variables. it was added to the style guide at some point

        - **Client ID**: Client ID from authentik.
        - **Client secret**: Client secret from authentik.
        - **Metadata address**: `https://authentik.company/application/o/<application_slug>/.well-known/openid-configuration`
        - **OIDC redirect behavior**: `Redirect GET`

>         - **Get claims from user info endpoint**: Checked.

Reviewer: ```suggestion
        - **Get claims from user info endpoint**: Select this option.
```

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings. Take note of the **slug** as it will be required later.
    - **Choose a Provider type**: select **SAML Provider** as the provider type.
    - **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
        - Temporarily set the **ACS URL** to `https://temp.temp`
        - Under **Advanced protocol settings**, set **Signing Certificate** to use any available certificate.
    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

### Download certificate file

1. Log in to authentik as an admin, and open the authentik Admin interface.

> 2. Navigate to **Applications** > **Providers** and click on the name of the provider that you created in the previous section (e.g. `Provider for bitwarden`).

Reviewer 2: ```suggestion
2. Navigate to **Applications** > **Providers** and click on the name of the provider that you created in the previous section (e.g. `Provider for Bitwarden`).
```

Unfortunately, we're gonna have to ruin the lowercase aesthetic vibe since the product name uses a capital letter

4. Under **Related objects** > **Download signing certificate**, click on **Download**. This downloaded file is your certificate file and it will be required in the next section.

## Bitwarden configuration

1. Log in to the [Bitwarden dashboard](https://vault.bitwarden.com/#/login) as an administrator (select `Accessing: bitwarden.eu` for EU based Bitwarden accounts). For self-hosted Bitwarden use `https://bitwarden.company/#/login`.
2. In the sidebar, navigate to **Admin Console** > **Settings** > **Single sign-on**, and enter the following settings:

>     - **Allow SSO authentication**: Checked.

Reviewer: ```suggestion
    - **Allow SSO authentication**: Select this option.
```

    - **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
    - **Choose a Provider type**: select **OAuth2/OpenID Connect** as the provider type.
    - **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
        - Note the **Client ID**, **Client Secret**, and **slug** values because they will be required later.
        - Set a `Strict` redirect URI to:
            - `https://sso.bitwarden.com/oidc-signin` for Non-EU based SaaS Bitwarden.
            - `https://sso.bitwarden.eu/oidc-signin` for EU based SaaS Bitwarden.
            - `https://bitwarden.company/oidc-signin` for self-hosted Bitwarden.
        - Select any available signing key.
        - Under **Advanced protocol settings**:
            - Remove `authentik default OAuth Mapping: OpenID 'profile'` from **Selected Scopes**.

>             - Add the scope mapping that you created to **Selected Scopes**.

Author: Do you prefer this?

>             - Add the scope mapping that you created to **Selected Scopes**.

Reviewer 2: looks perfaect
